wasm-rpc/src/lib.rs

```diff
@@ -304,6 +447,1012 @@ fn build_tree(node: &WitNode, nodes: &[WitNode]) -> Value {
     }
 }
 
+#[cfg(feature = "typeinfo")]
+#[cfg(not(feature = "stub"))]
+#[derive(Clone, Debug)]
+pub enum TypeAnnotatedValue {
+    Bool(bool),
+    S8(i8),
+    U8(u8),
+    S16(i16),
+    U16(u16),
+    S32(i32),
+    U32(u32),
+    S64(i64),
+    U64(u64),
+    F32(f32),
+    F64(f64),
+    Chr(char),
+    Str(String),
+    List {
+        typ: AnalysedType,
+        values: Vec<TypeAnnotatedValue>,
+    },
+    Tuple {
+        typ: Vec<AnalysedType>,
+        value: Vec<TypeAnnotatedValue>,
+    },
+    Record {
+        typ: Vec<(String, AnalysedType)>,
+        value: Vec<(String, TypeAnnotatedValue)>,
+    },
+    Flags {
+        typ: Vec<String>,
+        values: Vec<String>,
+    },
+    Variant {
+        typ: Vec<(String, Option<AnalysedType>)>,
+        case_name: String,
+        case_value: Option<Box<TypeAnnotatedValue>>,
+    },
+    Enum {
+        typ: Vec<String>,
+        value: String,
+    },
+    Option {
+        typ: AnalysedType,
+        value: Option<Box<TypeAnnotatedValue>>,
+    },
+    Result {
+        ok: Option<Box<AnalysedType>>,
+        error: Option<Box<AnalysedType>>,
+        value: Result<Option<Box<TypeAnnotatedValue>>, Option<Box<TypeAnnotatedValue>>>,
+    },
+    Handle {
+        id: AnalysedResourceId,
+        resource_mode: AnalysedResourceMode,
+        uri: Uri,
+        resource_id: u64,
+    },
+}
+
+#[cfg(feature = "typeinfo")]
+#[cfg(not(feature = "stub"))]
+impl TypeAnnotatedValue {
+    pub fn from_value(
+        val: Value,
```

Comment:
You are not really taking ownership of `val`, as you are cloning parts of it in several cases below. So I think it should be `&Value`.